### PR TITLE
fix(http): handle resource cleanup and prevent NPE during phase termination

### PR DESCRIPTION
**Pull request by diegolovison**
- Fixes a NullPointerException in HttpRequestContext where `this.request` is null when acquiring a connection.
- Stops the connection pool pulse loop when a phase is terminating by introducing `Session.tryTerminate()`.
- Prevents connection leaks during termination by clearing the waiting queue and releasing active connections back to the pool.

Fixes: 
- https://github.com/Hyperfoil/Hyperfoil/issues/643
- https://github.com/Hyperfoil/Hyperfoil/issues/642
- https://github.com/Hyperfoil/Hyperfoil/issues/638

Stacktrace before the fix
```
09:21:46,049 WARN  (epollEventLoopGroup-2-1) [i.h.h.c.Http1xConnection] Exception in Http1xConnection{/127.0.0.1:46166 -> localhost/127.0.0.1:41281, status=OPEN, size=0+1:[]}java.lang.NullPointerException: Cannot read field "session" because "this.request" is null
	at io.hyperfoil.http.steps.HttpRequestContext.accept(HttpRequestContext.java:37)
	at io.hyperfoil.http.connection.SharedConnectionPool.pulse(SharedConnectionPool.java:203)
	at io.hyperfoil.http.connection.SharedConnectionPool.handleNewConnection(SharedConnectionPool.java:325)
	at io.hyperfoil.http.connection.Http1xConnection.checkActivated(Http1xConnection.java:75)
	at io.hyperfoil.http.connection.Http1xConnection.channelActive(Http1xConnection.java:69)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:238)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelActive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelActive(DefaultChannelPipeline.java:1345)
```

### CODE REVIEW

File: test-suite/src/test/java/io/hyperfoil/scenario/WrkScenarioTest.java [hunk 1]
```diff
@@ -33,6 +33,16 @@ public class WrkScenarioTest extends BaseWrkBenchmarkTest {
 
    protected final Logger log = LogManager.getLogger(getClass());
 
+   @Test
+   public void wrk2SuperSlowServer() throws URISyntaxException {
```
Comment (franz1981) [inline test-suite/src/test/java/io/hyperfoil/scenario/WrkScenarioTest.java:37]: which output you have while you run this without your patch and with your patch?

Comment (diegolovison) [reply test-suite/src/test/java/io/hyperfoil/scenario/WrkScenarioTest.java:37]: If we run without the patch, it will produce NPE and fail.

Comment (franz1981) [reply test-suite/src/test/java/io/hyperfoil/scenario/WrkScenarioTest.java:37]: I will try reproduce it myself, but please in the description print the stack trace with the error too to help reviewers

Comment (diegolovison) [reply test-suite/src/test/java/io/hyperfoil/scenario/WrkScenarioTest.java:37]: Done

**Comment by franz1981** [comment]
Two critical issues with the termination mechanism:

1. shouldPulse is a plain boolean written from tryTerminate() (controller thread) but read from pulse() on the Netty event loop. No happens-before. Needs volatile or dispatching via executor().execute(...).
2. SharedConnectionPool is shared by all sessions on the same event loop. onSessionTryTerminate() sets pool-wide shouldPulse = false, and the next pulse() clears the entire waiting queue — including requests from sessions that are still active. The termination mechanism needs to be per-session, not per-pool.

Also: HttpRequestContext.accept() leaks the connection when request == null (should release it back to the pool).
The release() assert-to-if change and NPE null guard are correct individually.

**Comment by diegolovison** [comment]
@franz1981 please review again.

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java [hunk 2]
```diff
@@ -358,4 +391,12 @@ public void shutdown() {
          }
       });
    }
+
+   @Override
+   public void onSessionTryTerminate() {
+      if (!executor().inEventLoop()) {
+         executor().execute(this::onSessionTryTerminate);
```
Comment (franz1981) [inline http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:467]:
```suggestion
         executor().execute(this::onSessionTryTerminate);
         return; 
```

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java [hunk 3]
```diff
@@ -358,4 +391,12 @@ public void shutdown() {
...
+      }
+      this.shouldPulse = false;
```
Comment (franz1981) [inline http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:474]: The last line runs from controller thread too, do you need a `return`? see https://github.com/Hyperfoil/Hyperfoil/pull/816/changes#r3388597046

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java [hunk 4]
```diff
@@ -192,9 +195,39 @@ public EventLoop executor() {
 
    @Override
    public void pulse() {
+      assert executor().inEventLoop();
       if (trace) {
          log.trace("Pulse to {} ({} waiting)", authority, waiting.size());
       }
+      // session terminated, there is nothing that we can do
+      if (!shouldPulse) {
```
Comment (franz1981) [inline http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:239]: SharedConnectionPool is shared across all sessions on the same event loop. 
When any session terminates, `shouldPulse = false` on the shared pool, and `pulse()` removes ALL waiting consumers — because `isValid()` returns ready, which is false for every consumer in the waiting queue (ready is only set in `accept()`, which hasn't been called on waiting consumers). 
Requests from still-active sessions are discarded.

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java [hunk 5]
```diff
@@ -192,9 +195,39 @@ public EventLoop executor() {
...
+         }
+
+         blockedSessions.decrementUsed(removedCount);
+
+         if (blockedSessions.current() > 0) {
+            this.pulse();
+            return;
+         }
+
+         for (HttpConnection conn : connections) {
```
Comment (franz1981) [inline http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:255]: For connections with in-flight requests: this adds them to available now, and the response completion path adds them again later. 
ArrayDeque allows duplicates — same connection acquired by two consumers.

**Comment by franz1981** [comment]
The accept() null guard with connection release, and the release() assert-to-if change - is fine.

What needs redesign is the termination mechanism: it needs to be per-session, not per-pool. 
One approach: have pulse() check each waiting consumer's session termination status, rather than using a pool-wide flag.

**Review by franz1981 — CHANGES_REQUESTED** [review]

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java (hunk 2 above)
Comment (diegolovison) [reply http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:467]: Done

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java (hunk 3 above)
Comment (diegolovison) [reply http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:474]: Done

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java (hunk 4 above)
Comment (diegolovison) [reply http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:239]: > SharedConnectionPool is shared across all sessions on the same event loop.

Yes

> When any session terminates, shouldPulse = false on the shared pool

Yes, but the method is called only when `PhaseInstance::tryTerminate` is called and after all sessions have proceeded

> and pulse() removes ALL waiting consumers — because isValid() returns ready, which is false for every consumer in the waiting queue (ready is only set in accept(), which hasn't been called on waiting consumers).

Good catch. I removed the `isValid` and removed all from the waiting queue.

> Requests from still-active sessions are discarded.

The only edge case that I found was during `pulse`. Double-check `SessionImpl::run`. If an active session reaches that method with `PhaseInstance.Status.TERMINATING`, then it will stop and won't run.

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java (hunk 5 above)
Comment (diegolovison) [reply http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:255]: After `PhaseInstance.Status.TERMINATING`, the connection won't be added back. This PR also addresses https://github.com/Hyperfoil/Hyperfoil/issues/643

**Comment by diegolovison** [comment]
> What needs redesign is the termination mechanism: it needs to be per-session, not per-pool.
> One approach: have pulse() check each waiting consumer's session termination status, rather than using a pool-wide flag.

It was my initial approach. But this idea, from my findings, will require a big refactor.

@franz1981 the comments were addressed. Please review again.

**Comment by franz1981** [comment]
Two questions on the remaining design:

1. You said onSessionTryTerminate() is only called after all sessions have proceeded. Can you point to where this invariant is enforced? If SharedConnectionPool is shared across all sessions on the event loop, a single session's termination could trigger onSessionTryTerminate() while other sessions are still active.
2. You said connections won't be added back after TERMINATING status. Which code path prevents Http1xConnection.handleResponse() → pool.release() from adding a connection back to available during termination?

If both invariants hold, the current approach is correct. I just want to verify them.

**Comment by diegolovison** [comment]
> You said onSessionTryTerminate() is only called after all sessions have proceeded. Can you point to where this invariant is enforced? If SharedConnectionPool is shared across all sessions on the event loop, a single session's termination could trigger onSessionTryTerminate() while other sessions are still active.

After your question at https://github.com/Hyperfoil/Hyperfoil/pull/823#discussion_r3388582797 this PR requires https://github.com/Hyperfoil/Hyperfoil/pull/823 to be merged first. Then, I will need to update this PR again to address 

```java
for (int i = 0; i < sessionList.size(); i++) {
   Session session = sessionList.get(i);
   // add resources back
   session.tryTerminate();
}
```

> You said connections won't be added back after TERMINATING status. Which code path prevents Http1xConnection.handleResponse() → pool.release() from adding a connection back to available during termination?

It is describe at https://github.com/Hyperfoil/Hyperfoil/issues/643

File: http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java [hunk 6]
```diff
@@ -192,9 +194,26 @@ public EventLoop executor() {
...
       }
+      // session terminated, there is nothing that we can do
+      if (!shouldPulse) {
+
+         blockedSessions.decrementUsed(waiting.size());
+         waiting.clear();
+
+         assert blockedSessions.current() == 0;
+
+         for (HttpConnection conn : connections) {
```
Comment (franz1981) [inline http/src/main/java/io/hyperfoil/http/connection/SharedConnectionPool.java:255]: Another concern on this...
`release(conn, true, false)` adds the connection to available but does NOT decrement inFlight (because `afterRequest=false`). 
But `acquireNow()` incremented `inFlight` when the connection was acquired. 
If the connection was acquired for a consumer that's now in the cleared waiting queue, `inFlight` remains inflated.

Also: for connections with in-flight HTTP requests that complete after this cleanup, the response path calls `release(conn, true, true)`, adding the connection to `available` a second time (`ArrayDeque` allows duplicates). 

Your claim that "connections won't be added back after TERMINATING" needs to be verified: `Http1xConnection.handleResponse()` doesn't check phase status before calling `pool.release()`.

**Review by franz1981 — CHANGES_REQUESTED** [review]

**Review by franz1981 — COMMENTED** [review]
The test has two issues:

1. It checks `inFlightCount` inside the [completion handler](https://github.com/Hyperfoil/Hyperfoil/pull/816/files#diff-8d9e82e8f5ac0b7aef22aba9c4e53f2de9f62b9d63f5c2b42c36c2e6f0d62f25R78-R101), which fires [before `releasePoolAndPulse()` decrements `inFlight`](https://github.com/Hyperfoil/Hyperfoil/blob/76625d30/http/src/main/java/io/hyperfoil/http/connection/Http1xResponseHandler.java#L544-L565). The `inFlight == 1` it sees is the normal state at that point in the response pipeline — not evidence of a leak.

2. [`session.tryTerminate()` at line 113](https://github.com/Hyperfoil/Hyperfoil/pull/816/files#diff-8d9e82e8f5ac0b7aef22aba9c4e53f2de9f62b9d63f5c2b42c36c2e6f0d62f25R113) never sets `shouldPulse = false` because [`initForTesting` registers an `HttpDestinationTableImpl` with empty maps](https://github.com/Hyperfoil/Hyperfoil/blob/76625d30/http/src/main/java/io/hyperfoil/http/HttpRunData.java#L88) — the `!shouldPulse` code path is never entered. The test doesn't exercise the race condition it describes.

**Comment by diegolovison** [comment]
@franz1981 Please review again

**Comment by franz1981** [comment]
`HttpRequestContext.accept()` with `request == null` leaks pool-level `inFlight` by 1.

`cancelAcquire()` undoes `conn.onAcquire()` (aboutToSend--) but `release(conn, true, false)` passes `afterRequest=false`, so `inFlight.decrementUsed()` is never called to undo the `acquireNow()` increment.

Reproducer test:

```java
@Test
public void testAcceptNullRequestLeaksPoolInFlight(VertxTestContext ctx) {
   var checkpoint = ctx.checkpoint();

   getClientPool(ctx).onComplete(result -> {
      if (result.failed()) return;
      HttpClientPool client = result.result();

      SharedConnectionPool pool = (SharedConnectionPool) client.next();
      pool.executor().execute(() -> {
         Session session = SessionFactory.forTesting();
         HttpRunData.initForTesting(session);

         // Acquire the only connection so the pool is exhausted
         pool.acquire(false, (HttpConnection firstConn) -> {

            // Put a consumer in the waiting queue that mimics
            // HttpRequestContext.accept() when request==null
            pool.acquire(false, (HttpConnection conn) -> {
               if (conn != null && conn.pool() != null) {
                  conn.cancelAcquire();
                  conn.pool().release(conn, true, false);
               }

               ctx.verify(() -> {
                  assertThat(pool.inFlightCount())
                        .as("pool inFlight should be 0 — cancelAcquire undoes conn-level but not pool-level")
                        .isEqualTo(0);  // FAILS: actual is 1
               });
               checkpoint.flag();
            });

            // Release firstConn to trigger pulse() which pops the waiting consumer
            firstConn.cancelAcquire();
            pool.release(firstConn, true, true);
            pool.pulse();
         });
      });
   });
}
```

Result: `expected: 0 but was: 1`

Fix: add `inFlight.decrementUsed()` in the null-request path of `accept()`, same as the [orphaned-acquisition drain](https://github.com/Hyperfoil/Hyperfoil/pull/816/files#diff-2a969a23f7a3ced3e4e9b44e2e6f2e88a6b02dbf91b9b34bcc7bc998d19acbb9R275-R278).

**Comment by diegolovison** [comment]
@franz1981 ready for review


**A quick note on the test tweak and a minor edge case it highlighted:**

I had to add `firstConn.attach(pool);` in the test because our mock doesn't call `request.send()` (which is normally where a connection gets linked to its pool).

Interestingly, this actually exposes a non-obvious edge case in the production code: if a **brand new** connection (which hasn't been attached to a pool yet) hits this `null` request abort *before* `send()` is called, `connection.pool()` will evaluate to null. This means it would bypass the `cancelAcquire` block entirely.

This patch successfully fixes the leak for reused connections (which covers the vast majority of cases), but it suggests that in a future PR, we might want to move `connection.attach(pool)` upstream into the connection creation logic itself, rather than waiting for `HttpRequest.send()`, so connections always know their pool from birth."

**Review by franz1981 — APPROVED** [review]
Fix is correct. 138/138 http tests pass. The `cancelAcquire(HttpConnection)` method cleanly consolidates `conn.cancelAcquire()` + `inFlight.decrementUsed()` + `release(conn, true, false)` and is used from both the drain path and `HttpRequestContext.accept()`.

Agree on the brand-new-connection edge case — `connection.pool()` is null until `HttpRequest.send()` calls `connection.attach(pool)`, so a never-used connection would skip the `cancelAcquire` block. Fine to address in a follow-up (moving `attach(pool)` upstream into connection creation).

LGTM.

**Comment by franz1981** [comment]
Well done @diegolovison ! this was a quite deep rabbit hole!

**Comment by diegolovison** [comment]
@franz1981 thanks for the review and for pushing this forward!
